docs: Create Buildkite Integration document (#865)
platform-platform/monorepo: docs/17_buildkite_ci_integration.md

  2. Select your organization in the dropdown on the top bar.
  3. Press the `New Pipeline` (a plus **+**) button above the list of organization pipelines.
  4. Name a new pipeline as `Sync Pipeline` and enter the Git repository URL for a project you would like to export build data.
  5. Press the `Create Pipeline` button and skip the `Webhook setup`.

Once a new pipeline is created, let's configure it to use the pipeline YAML file from the remote. Consider the next steps:

  1. Select a new `Sync Pipeline` on the same page with your organization's pipelines.
  2. Press the `Edit Steps` button on the navigation bar at the top. 
  3. Press the `Convert to YAML Steps` button.
  4. Replace the contents of the opened editor with the following (**.buildkite/path/to/** is a path to the sync pipeline YAML file in your Git repository):
  ```yaml
  steps:
    - command: "buildkite-agent pipeline upload .buildkite/path/to/sync_pipeline.yml"
      label: "Upload Sync Pipeline"
  ```

Finally, let's create a pipeline YAML file for the **Sync Pipeline**. In the connected repository create a new file `.buildkite/path/to/sync_pipeline.yml` with the following contents:

>  
>   1. In a connected repository, create or browse to `.buildkite/pipelines` folder.
>   2. Create `sync_pipeline.yml` file.
>   3. Add the following contents to the file:

Reviewer: ```suggestion
```

Please note that `Sync Pipeline` uses the configuration file and contains a path to the directory where this file is located. So we should create a config under the same directory as specified in the `Sync Pipeline`. Also, its name should match the name from the pipeline. 
Once the file `path/to/config/directory/integration_config.yml` is created, add the following lines:
```yaml
source:
  buildkite:
    access_token: $BUILDKITE_TOKEN
    pipeline_slug: $PIPELINE_SLUG
    organization_slug: <YOUR_ORGANIZATION_SLUG>
destination:
  firestore:
    firebase_project_id: <YOUR_FIREBASE_PROJECT_ID>
    firebase_user_email: $WEB_APP_USER_EMAIL
    firebase_user_pass: $WEB_APP_USER_PASSWORD
    firebase_public_api_key: $CI_INTEGRATIONS_FIREBASE_API_KEY
    metrics_project_id: $METRICS_PROJECT_ID
```

First, let's consider these types of values in the config file:

> 
> There are three types of config values depending on how they are exposed:

Reviewer: ```suggestion
```

>   -`Secret` is a value that must be stored securely using the [Buildkite secrets in environment hooks](https://buildkite.com/docs/pipelines/secrets#storing-secrets-in-environment-hooks).

Reviewer: ```suggestion
  - `Secret` is a value that must be stored securely using the [Buildkite secrets in environment hooks](https://buildkite.com/docs/pipelines/secrets#storing-secrets-in-environment-hooks).
```

  - `Pipeline` is a pipeline environment variable value that is set from a trigger step environment. Read more about this in a [Triggering `Sync Pipeline` section](#Triggering-Sync-Pipeline).
  - `Public` is a value that can be defined directly in the config file and is **specific to you**.

>   Using the above definitions, consider the following table with config variables:

Reviewer: ```suggestion
  Using the above definitions, consider the following table with configuration keys:
```

 | `firebase_user_email` | `Secret` | A Firebase user email to authorize with Metrics |
 | `firebase_user_pass` | `Secret` | A Firebase user password to authorize with Metrics | 
 | `firebase_public_api_key` | `Public` / `Secret` (if `Firebase key protection` is enabled) | A Firebase API key |
 | `metrics_project_id` | `Pipeline` | A Firestore Metrics project identifier |

As the Sync Pipeline is created and the configuration file for CI integration is ready to use, it's time to configure the pipeline, that we want to export build data for, to trigger synchronization.

## Triggering Sync Pipeline

The main idea of auto-exporting build data is asynchronous triggering the `Sync Pipeline` from the pipeline we want to synchronize. Thus, when the pipeline finishes all of its general steps, it triggers a `Sync Pipeline` to start and provides variables with information about itself. After the async trigger, the pipeline completes, and its results are exported to the Metrics project. The trigger step must depend on the very last pipeline-related step. In other words, the trigger step must be the latest.

Let's consider the step described:

> 

Reviewer: ```suggestion
```

```yaml
  - label: "Trigger sync pipeline"
    trigger: sync-pipeline
    depends_on: "step_key"
    allow_dependency_failure: true
    async: true
    build:
      env:
        PIPELINE_SLUG: "${BUILDKITE_PIPELINE_SLUG}"
        METRICS_PROJECT_ID: "metrics_project_id"
```

> The `allow_dependency_failure` flag allows running the trigger even if the step with the key `step_key` finishes with an error. We strongly recommend setting the `depends_on` value to the step key that finalizes your build (waits on any other steps to finish). 

Reviewer: ```suggestion
The `allow_dependency_failure` flag allows running the trigger even if the step with the key `step_key` finishes with an error. We strongly recommend setting the `depends_on` value to the key of a step that finalizes your build (waits on any other steps to finish). 
```

> The `build` section provides the `Sync Pipeline` with environment variables containing the information about what pipeline triggers exporting (`PIPELINE_SLUG` variable with `BUILDKITE_PIPELINE_SLUG` value) and what project in the database matches this pipeline (`METRICS_PROJECT_ID` variable with the **specific to you** value).

Reviewer: ```suggestion
The `build` section provides the `Sync Pipeline` with environment variables containing the information about what pipeline is triggering export (`PIPELINE_SLUG` variable with `BUILDKITE_PIPELINE_SLUG` value) and what project in the database matches this pipeline (`METRICS_PROJECT_ID` variable with the **specific to you** value).
```

> 

Reviewer: ```suggestion

To read more about the `trigger` step, consider the Buildkite [Trigger Step](https://buildkite.com/docs/pipelines/trigger-step) documentation. Also, the [Managing Step Dependencies](https://buildkite.com/docs/pipelines/dependencies) documentation can help you to understand how to make your trigger step to be the latest.

```

> Well done! Once the trigger step is configured, the auto-exports are configured and ready to use.

Reviewer: ```suggestion
Well done! Once the trigger step is configured, the auto-exporting is ready to use.
```

Here's an overview of each field in this step:

- `trigger` is a command that triggers another pipeline with the given `pipeline_slug`(in our case - `sync-pipeline`).
- `depends_on` - is a command, that indicates that this step will not run until the `step_key` step has been completed. You need to replace the `step_key` with the previous step's key this step would depend on.
- `allow_dependency_failure` - is a command, that forces this step to run even if the previous step fails.
- `async` - is a command, that must be set to `true` and forces this step to finish before the triggered pipeline finishes.

Also, you need to define the `environment variables` passed to the `Trigger sync pipeline` from the current pipeline:
 - `PIPELINE_SLUG` - is a unique identifier of the current pipeline. This variable is pre-defined in the pipeline, so you don't need to change it.
 - `METRICS_PROJECT_ID` - is a Firestore Metrics project identifier - you need to replace it with your id.

# API

> What will the proposed API look like?

> Once we've configured the Buildkite integration, let's consider the sequence diagram that will explain the main relationships between the different pipelines on the example with the `AwesomePipeline`:

Reviewer: ```suggestion
The algorithm for auto-exporting builds data in Buildkite uses three main components: the pipeline we want to export, the `CI integration` tool, and the `Sync Pipeline`. The following sequence diagram demonstrates how these components behave together and how the exporting happens (assume the pipeline to export as `Awesome Pipeline`):
```
